### PR TITLE
feat: Support for RS2 Downsampler

### PR DESCRIPTION
**Pull request by MaxiBoether**
This implements the random selection from the [RS2 paper](https://openreview.net/pdf?id=JnRStoIuTe) minus the learning rate scheduling adjustments.

Note that it is a bit suboptimal to use the downsampling infrastructure here (#466). We might want to think about making the selector a bit more dynamic, but for now, this will suffice to run experiments. #462 should be merged before this is reviewed.

### CODE REVIEW

File: modyn/config/schema/sampling/downsampling_config.py [hunk 1]
```diff
@@ -121,6 +121,33 @@ class RHOLossDownsamplingConfig(BaseDownsamplingConfig):
...
+    @classmethod
+    def sample_then_batch_must_be_true(cls, v: bool) -> bool:
+        if not v:
+            raise ValueError("sample_then_batch must be set to True for this config.")
+        return v
+
+    @field_validator("period")
+    @classmethod
+    def only_support_period_one(cls, v: int) -> int:
+        if v != 0:
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/config/schema/sampling/downsampling_config.py]: solved

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+        # Epoch 2
+        downsampler.inform_samples(sample_ids, data, target)
+        epoch2_ids, weights = downsampler.select_points()
+        sampled_data, sampled_target = get_tensors_subset(epoch2_ids, data, target, sample_ids)
+
+        assert len(set(epoch2_ids)) == 5
+        assert weights.shape == (5,)
+        assert all(idx in sample_ids for idx in epoch2_ids)
+        assert not any(idx in epoch1_ids for idx in epoch2_ids)  # No overlap across epochs
```
Comment (XianzheMa) [inline modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py:126]: How does this line absolutely pass? For each epoch we reshuffle the sample ids. Even if now we choose the second subset, the ids in the second subset can still come from the first subset in the previous epoch right? (because there is not a definite relation between the first subset in the first epoch and the second subset in the second epoch, as we shuffle the entire sample ids in between)

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    downsampler = RemoteRS2Downsampling(
+        pipeline_id, trigger_id, batch_size, params_from_selector, per_sample_loss, device
+    )
+    with torch.inference_mode(mode=(not downsampler.requires_grad)):
+        sample_ids = list(range(10))
+        data = torch.randn(10, 10)
+        target = torch.randint(0, 10, (10,))
+
+        for _ in range(3):
+            downsampler.inform_samples(sample_ids, data, target)
```
Comment (XianzheMa) [inline modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py:79]: Each call to `inform_samples` should be provided with a different set of `sample_ids`

**Review by XianzheMa — CHANGES_REQUESTED** [review]

File: modyn/config/schema/sampling/downsampling_config.py [hunk 4]
```diff
@@ -121,6 +121,33 @@ class RHOLossDownsamplingConfig(BaseDownsamplingConfig):
...
+            # RS2 requires us to resample every epoch.
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/config/schema/sampling/downsampling_config.py:146]: But what if users set this to something else, then? Will it still fail?

File: modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsampling_strategy.py [hunk 5]
```diff
@@ -62,6 +62,16 @@ def __init__(
...
 
+        # Some methods might not need information from forward pass (e.g. completely random)
+        # Most do (definition), hence we default to True
+        # We might want to refactor those downsamplers to presamplers and support some
+        # adaptivity at the selector, but for now we allow random downsamplers mostly
+        # to support RS2.
+        self.forward_required = True
+
+        # Some methods might only support StB, not BtS.
+        self.supports_bts = True
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsampling_strategy.py:73]: I think we should have this. The default is bts, and if for some reason some downsampling policy does not support it, it can override the field.

The problem is that the pydantic validation happens on the user side of things. But if we at some future point decide to use the downsampling code in other contexts (similar to how we use the training code now in RHO-LOSS), then suddenly user validations do not hold anymore. Since BTS RS2 is undefined behavior, I'd rather fail here then have silent errors that one does not notice

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py [hunk 6]
```diff
@@ -374,6 +374,8 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
             self._info(f"Training will stop when the number of samples to pass reaches {self.num_samples_to_pass}.")
 
         if self._downsampling_mode == DownsamplingMode.BATCH_THEN_SAMPLE:
+            # assertion since model validation by pydantic should catch this.
+            assert self._downsampler.supports_bts, "The downsampler does not support batch then sample"
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:378]: see my other comment on why I'd prefer to have this

File: modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        # We only need to collect the sample information once
+        if self._first_epoch:
+            self._all_sample_ids.extend(sample_ids)
+
+    def _epoch_step_wr(self, target_size: int) -> None:
+        self._subsets = [self._all_sample_ids[:target_size]]
+        self._current_subset = 0
+
+    def _epoch_step_r(self, target_size: int) -> None:
+        self._max_subset = len(self._all_sample_ids) // target_size
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py:57]: True, will fix

File: modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        self._current_subset += 1
+        if self._current_subset >= self._max_subset or len(self._subsets) == 0:
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py:59]: In the very first round, this will be the case. We need some form of flag to indicate reshuffling, either after we passed through all subsets or have not done anything yet

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py (hunk 2 above)
Comment (MaxiBoether) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py:126]: Sorry, I don't understand your question. In RS2, we generate a new subset each epoch (= round in the paper). Now, with replacement, we just always get some x% subset of the IDs. When we do without replacement, we have to iterate through the entire dataset (minus some leftover depending on the size), and then restart. Imagine we have these IDs

1 2 3 4 5

now we shuffle

2 1 4 5 3

now we partition it into subsets

[2 1] [4 5] 3

first epoch will return 2 1, second epoch will return 4 5

then we will have to reshuffle

3 2 1 5 4

partition again

[3 2] [1 5] 4

third epoch will return 3 2, fourth will return 1 5

Does this answer your question?

File: modyn/config/schema/sampling/downsampling_config.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/config/schema/sampling/downsampling_config.py:146]: Ok, tested it. Thanks, will proceed with your approach

[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/config/schema/sampling/downsampling_config.py:146]: Changed

File: modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py:57]: Fixed

File: modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import torch
+from modyn.trainer_server.internal.trainer.remote_downsamplers.abstract_remote_downsampling_strategy import (
+    AbstractRemoteDownsamplingStrategy,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class RemoteRS2Downsampling(AbstractRemoteDownsamplingStrategy):
+    """
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py:14]: Done

**Comment by MaxiBoether** [comment]
Answered your comments and addressed them where possible for now :)

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py (hunk 3 above)
Comment (XianzheMa) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py:79]: Yeah it does not make a difference here, as we just test the shape. But naturally they should be different because,

In `sample_and_batch`. In the `pytorch_trainer.py`, we first iterate over the dataloader and keep informing each batch in `_iterate_dataloader_and_compute_scores` https://github.com/eth-easl/modyn/blob/ac66eafc55c9089f938436e6255062df805bde9a/modyn/trainer_server/internal/trainer/pytorch_trainer.py#L836

the `sample_ids` come from the dataloader and should be naturally distinct right? (they are keys of the samples)

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py (hunk 2 above)
Comment (XianzheMa) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py:126]: Ah I see.I think this line confuses me https://github.com/eth-easl/modyn/blob/240dbf49de69b27821ea39c36867aae0e3bb66bb/modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py#L65

Actually, in the `NoReplacement` case, we only need to reshuffle at https://github.com/eth-easl/modyn/blob/240dbf49de69b27821ea39c36867aae0e3bb66bb/modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py#L60 right? Can you move the shuffle statement to each `step` method?

File: modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (XianzheMa) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py]: Thanks! that explains it

**Review by XianzheMa — CHANGES_REQUESTED** [review]

File: modyn/selector/internal/selector_strategies/downsampling_strategies/rs2_downsampling_strategy.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+class RS2DownsamplingStrategy(AbstractDownsamplingStrategy):
+    def __init__(
+        self,
+        downsampling_config: RS2DownsamplingConfig,
+        modyn_config: dict,
+        pipeline_id: int,
+        maximum_keys_in_memory: int,
+    ):
+        super().__init__(downsampling_config, modyn_config, pipeline_id, maximum_keys_in_memory)
+        self.remote_downsampling_strategy_name = "RS2DownsamplingStrategy"
```
[the remark this answers is not in the document]
Comment (XianzheMa) [reply modyn/selector/internal/selector_strategies/downsampling_strategies/rs2_downsampling_strategy.py:14]: Also the name `self.remote_downsampling_strategy_name` should match the remote class name. That's how the remote class is correctly selected and initialized. Relevant code is in `pytorch_trainer.py` feel free to take a look at how it is initialized.

**Review by XianzheMa — APPROVED** [review]
There are still some points to address, but after that, feel free to merge🚀, thanks a lot!

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py (hunk 2 above)
Comment (MaxiBoether) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py:126]: Sure. I am not opinionated on where we shuffle (I thought it's not an expensive operation so I can just unify w/wo replacement) but I moved it into the step functions

File: modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsampling_strategy.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsampling_strategy.py:73]: I understand your concern, but I like [defensive programming](https://en.wikipedia.org/wiki/Defensive_programming). I think it's a tradeoff between complexity and validating internal data. It's like using `>=` as a loop exit condition when actually you think only `==` is going to happen - are you _really_ sure that you variable will not for whatever reason increase by 2 within a single iteration?

I'd add the bts flag when the time comes that such a downsampler exists :D I am not saying we should add lots of assertions that are unnecessary, but where we do not have complete trust, I'd add them. The trainer is receiving input that goes via gRPC and whatnot - we have no guarantee (from the trainer perspective) that the input we got is validated. We asssume only the supervisor is going to send the trainer data here but who knows what will happen, anybody could send a serialized gRPC request.

TL;DR I would like to keep it

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:378]: I don't get that yet - the if only checks the flag. For reasons outlined above (see new comment), we cannot trust the input here. And I would not silently not downsample in case we have invalid input. We should fail when we get invalid input

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py (hunk 3 above)
Comment (MaxiBoether) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py:79]: No, they are not. What differs is the model output (on which true downsamplers sample), but the list of samples is always the same, since the trigger training set from the selector does not change between epochs. Since RS2 only relies on the IDs, it should not matter. The IDs will in all cases be identical across epochs.

File: modyn/selector/internal/selector_strategies/downsampling_strategies/rs2_downsampling_strategy.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/selector/internal/selector_strategies/downsampling_strategies/rs2_downsampling_strategy.py:14]: Thanks for catching this. I agree we should somehow integration test this, but this is a bigger task since it tests the interaction of two components. Let's do that separately

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    downsampler.inform_samples(sample_ids, forward_output, target)
+
+    assert downsampler._all_sample_ids == sample_ids
+    downsampler.inform_samples(sample_ids, forward_output, target)
+    assert downsampler._all_sample_ids == 2 * sample_ids
+    # Now it should not change anymore
+    downsampler.select_points()
+    downsampler.inform_samples(sample_ids, forward_output, target)
+    assert set(downsampler._all_sample_ids) == set(sample_ids)
+    assert len(downsampler._all_sample_ids) == 2 * len(sample_ids)
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py:59]: Not sure if this has been addressed by my explanation. `sample_id`s are consistent across epochs

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
+import torch
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py]: done!

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py (hunk 3 above)
Comment (XianzheMa) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py]: I think we have a mistake. I am saying the consecutive calls to `inform_samples` within two `select_points` calls should contain different sample ids.

I copy the code of `_iterate_dataloader_and_compute_scores` here:
```
        for batch_number, batch in enumerate(dataloader):
            self.update_queue(AvailableQueues.DOWNSAMPLING, batch_number, number_of_samples, training_active=False)

            sample_ids, target, data = self.preprocess_batch(batch)
            number_of_samples += len(sample_ids)

            with torch.inference_mode(mode=(not self._downsampler.requires_grad)):
                with torch.autocast(self._device_type, enabled=self._amp):
                    # compute the scores and accumulate them
                    model_output = self._model.model(data)
                    embeddings = self.get_embeddings_if_recorded()
                    self._downsampler.inform_samples(sample_ids, model_output, target, embeddings)
```

You see. We load batch from the dataloader. The first batch must have different sample ids than the second batch's sample ids.

I am not talking about sample ids across epochs. Those definitely do not change.

Comment (XianzheMa) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py]: i.e. if we have
```
downsampler.inform_samples(...)
downsampler.select_points(...)
downsampler.inform_samples(...)
```
Then the first inform_samples call can have the same sample ids as the second inform_samples.

But when we do 

```
downsampler.inform_samples(...)
downsampler.inform_samples(...)
downsampler.select_points(...)
downsampler.inform_samples(...)
downsampler.inform_samples(...)
```
Suppose the whole dataset contains two batches. Then the first two `inform_samples` calls should contain different `sample_ids`.

Comment (XianzheMa) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py]: Anyway, I think it does not really make a difference here to use different sample ids. But I still do think consecutive `inform_samples` calls (without `select_points` call in the middle) should contain distinct sample ids.

**Comment by XianzheMa** [comment]
Feel free to merge the PR! Thanks for the further explanation!!!!

File: modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py (hunk 3 above)
Comment (MaxiBoether) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py]: I understand your point and agree with your description, but I still don't understand why you are suggesting it here :D The code is this

```
    with torch.inference_mode(mode=(not downsampler.requires_grad)):
        sample_ids = list(range(10))
        data = torch.randn(10, 10)
        target = torch.randint(0, 10, (10,))

        for _ in range(3):
            downsampler.inform_samples(sample_ids, data, target)
            selected_ids, weights = downsampler.select_points()
```

so the loop is the epoch loop (!). Since `sample_ids = list(range(10))` we don't have duplicate samples in the same epoch and consistent samples across epochs. This is exactly like you describe. I am not sure if I am missing something or you just confused this loop with something else. I am merging this for now and happy to do a follow up PR in case I am missing something here

Comment (MaxiBoether) [reply modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py]: " this unit test, we only keep calling inform_samples(...) without calling select_points(...),"

i dont get it. isn't it directly below :D?
